### PR TITLE
search: add pattern type for new keyword search

### PR DESCRIPTION
**Pull request by stefanhengl**
This adds a new pattern type "newStandardRC1" which replaces "standard" if the feature flag "search-new-keyword" is set. Instead of calling Zoekt's parser, we concatenate literals with an `AND` operator, which should automatically provide 100% compatibility with our current search syntax. Previously we didn't support Sourcegraph specific parts of the query language, such as inline regexp, in our prototype.

To switch back and forth between "standard" and "newStandardRC1", I adapted the smart search selector like this (only visible if "search-new-keyword" is set) 

image

## Test plan
- I copied many of the tests for "standard" and adapted them for "newStandardRC1". Check out the new golden files and compare them to the corresponding golden files for "standard".
- manual testing
  -  tried out queries, such as `r:zoekt query parse`, and verified the results
  - mixed literals with regexp and verified that the results make sense.

### CODE REVIEW

**Comment by stefanhengl** [comment]
@fkling tagging you for the client changes.

File: internal/search/job/jobutil/job.go [hunk 1]
```diff
@@ -840,7 +840,7 @@ func (b *jobBuilder) newZoektSearch(typ search.IndexedRequestType) (job.Job, err
 		FileMatchLimit: b.fileMatchLimit,
 		Select:         b.selector,
 		Features:       *b.features,
-		KeywordScoring: b.patternType == query.SearchTypeKeyword,
+		PatternType:    b.patternType,
```
Comment (keegancsmith) [inline internal/search/job/jobutil/job.go:843]: intentional to remove KeywordScoring?

File: internal/search/client/client.go [hunk 2]
```diff
@@ -231,6 +231,8 @@ func SearchTypeFromString(patternType string) (query.SearchType, error) {
 		return query.SearchTypeLucky, nil
 	case "keyword":
 		return query.SearchTypeKeyword, nil
+	case "standardv2":
```
Comment (keegancsmith) [inline internal/search/client/client.go:234]: I'm wondering if while this is experimental we should include that in the wording. Less surpsing then when we rename/remove it.

File: internal/search/client/client.go [hunk 3]
```diff
@@ -252,8 +254,10 @@ func detectSearchType(version string, patternType *string) (query.SearchType, er
 			searchType = query.SearchTypeLiteral
 		case "V3":
 			searchType = query.SearchTypeStandard
+		case "V4":
```
Comment (keegancsmith) [inline internal/search/client/client.go:257]: does version have to look like v{num}. Could we do something like V4-rc1 or V4-exp or something?

**Review by keegancsmith — APPROVED** [review]
backend LGTM. Main feedback is using an experimental marker in what we marshal over the wire for search type and version.

File: internal/search/client/client.go (hunk 3 above)
Comment (stefanhengl) [reply internal/search/client/client.go:257]: Sure, that makes sense to me.

**Review by camdencheek — APPROVED** [review]
LGTM! Exciting stuff.

Unrelated to this PR, but commenting here so I don't forget: do we have plans to push down and/or support to `searcher`? If we make this search mode default, it will dramatically increase the number of AND queries, which are slow and unstable for unindexed search because searcher cannot handle them natively, so we do intersection/union in `frontend`.

**Comment by stefanhengl** [comment]
> LGTM! Exciting stuff.
> 
> Unrelated to this PR, but commenting here so I don't forget: do we have plans to push down and/or support to `searcher`? If we make this search mode default, it will dramatically increase the number of AND queries, which are slow and unstable for unindexed search because searcher cannot handle them natively, so we do intersection/union in `frontend`.

Good point. We definitely have to look into that. I will put it on our board. Thanks for commenting here.

**Pull request merged**